README.md
# sbr-merkle-patricia-tree

This is an implementation of the modified merkle patricia tree as specified in the [Ethereum Yellow Paper](http://gavwood.com/Paper.pdf):

> The modified Merkle Patricia tree (trie) provides a persistent data structure to map between arbitrary-length binary data (byte arrays). It is defined in terms of a mutable data structure to map between 256-bit binary fragments and arbitrary-length binary data. The core of the trie, and its sole requirement in terms of the protocol specification is to provide a single 32-byte value that identifies a given set of key-value pairs.

The only backing store supported is LevelDB through the `levelup` module.

# INSTALL

`npm install https://github.com/siliconswampio/sbr-merkle-patricia-tree.git`

# USAGE

There are 3 variants of the tree implemented in this library, namely: `BaseTrie`, `CheckpointTrie` and `SecureTrie`. `CheckpointTrie` adds checkpointing functionality to the `BaseTrie` with the methods `checkpoint`, `commit` and `revert`. `SecureTrie` extends `CheckpointTrie` and is the most suitable variant for Ethereum applications. It stores values under the `keccak256` hash of their keys.

## Initialization and Basic Usage

```typescript
import level from 'level'
import { BaseTrie as Trie } from 'sbr-merkle-patricia-tree'

const db = level('./testdb')
const trie = new Trie(db)

async function test() {
  await trie.put(Buffer.from('test'), Buffer.from('one'))
  const value = await trie.get(Buffer.from('test'))
  console.log(value.toString()) // 'one'
}

test()
```

## Merkle Proofs

```typescript
const trie = new Trie()

async function test() {
  await trie.put(Buffer.from('test'), Buffer.from('one'))
  const proof = await Trie.createProof(trie, Buffer.from('test'))
  const value = await Trie.verifyProof(trie.root, Buffer.from('test'), proof)
  console.log(value.toString()) // 'one'
}

test()
```

## Read stream on Geth DB

```typescript
import level from 'level'
import { SecureTrie as Trie } from 'sbr-merkle-patricia-tree'

const db = level('YOUR_PATH_TO_THE_GETH_CHAIN_DB')
// Set stateRoot to block #222
const stateRoot = '0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544'
// Convert the state root to a Buffer (strip the 0x prefix)
const stateRootBuffer = Buffer.from(stateRoot.slice(2), 'hex')
// Initialize trie
const trie = new Trie(db, stateRootBuffer)

trie
  .createReadStream()
  .on('data', console.log)
  .on('end', () => {
    console.log('End.')
  })
```

## Read Account State including Storage from Geth DB

```typescript
import level from 'level'
import rlp from 'rlp'
import { BN, bufferToHex } from 'sbr-util'
import Account from 'ethereumjs-account'
import { SecureTrie as Trie } from 'sbr-merkle-patricia-tree'

const stateRoot = 'STATE_ROOT_OF_A_BLOCK'

const db = level('YOUR_PATH_TO_THE_GETH_CHAINDATA_FOLDER')
const trie = new Trie(db, stateRoot)

const address = 'AN_ETHEREUM_ACCOUNT_ADDRESS'

async function test() {
  const data = await trie.get(address)
  const acc = new Account(data)

  console.log('-------State-------')
  console.log(`nonce: ${new BN(acc.nonce)}`)
  console.log(`balance in wei: ${new BN(acc.balance)}`)
  console.log(`storageRoot: ${bufferToHex(acc.stateRoot)}`)
  console.log(`codeHash: ${bufferToHex(acc.codeHash)}`)

  let storageTrie = trie.copy()
  storageTrie.root = acc.stateRoot

  console.log('------Storage------')
  const stream = storageTrie.createReadStream()
  stream
    .on('data', (data) => {
      console.log(`key: ${bufferToHex(data.key)}`)
      console.log(`Value: ${bufferToHex(rlp.decode(data.value))}`)
    })
    .on('end', () => {
      console.log('Finished reading storage.')
    })
}

test()
```

Additional examples with detailed explanations are available [here](https://github.com/gabrocheleau/merkle-patricia-trees-examples).

# API

[Documentation](./docs/README.md)

# TESTING

`npm test`

# BENCHMARKS

There are two simple **benchmarks** in the `benchmarks` folder:

- `random.ts` runs random `PUT` operations on the tree.
- `checkpointing.ts` runs checkpoints and commits between `PUT` operations.

A third benchmark using mainnet data to simulate real load is also under consideration.

Benchmarks can be run with:

```shell
npm run benchmarks
```

To run a **profiler** on the `random.ts` benchmark and generate a flamegraph with [0x](https://github.com/davidmarkclements/0x) you can use:

```shell
npm run profiling
```

0x processes the stacks and generates a profile folder (`<pid>.0x`) containing [`flamegraph.html`](https://github.com/davidmarkclements/0x/blob/master/docs/ui.md).

# REFERENCES

- Wiki
  - [Ethereum Trie Specification](https://github.com/ethereum/wiki/wiki/Patricia-Tree)
- Blog posts
  - [Ethereum's Merkle Patricia Trees - An Interactive JavaScript Tutorial](https://rockwaterweb.com/ethereum-merkle-patricia-trees-javascript-tutorial/)
  - [Merkling in Ethereum](https://blog.ethereum.org/2015/11/15/merkling-in-ethereum/)
  - [Understanding the Ethereum Trie](https://easythereentropy.wordpress.com/2014/06/04/understanding-the-ethereum-trie/). Worth a read, but the Python libraries are outdated.
- Videos
  - [Trie and Patricia Trie Overview](https://www.youtube.com/watch?v=jXAHLqQthKw&t=26s)

# EthereumJS

See our organizational [documentation](https://ethereumjs.readthedocs.io) for an introduction to `EthereumJS` as well as information on current standards and best practices.

If you want to join for work or do improvements on the libraries have a look at our [contribution guidelines](https://ethereumjs.readthedocs.io/en/latest/contributing.html).

# LICENSE

[MPL-2.0](<https://tldrlegal.com/license/mozilla-public-license-2.0-(mpl-2)>)

[discord-badge]: https://img.shields.io/static/v1?logo=discord&label=discord&message=Join&color=blue
[discord-link]: https://discord.gg/TNwARpR
[trie-npm-badge]: https://img.shields.io/npm/v/merkle-patricia-tree.svg
[trie-npm-link]: https://www.npmjs.com/package/merkle-patricia-tree
[trie-issues-badge]: https://img.shields.io/github/issues/ethereumjs/ethereumjs-monorepo/package:%20trie?label=issues
[trie-issues-link]: https://github.com/ethereumjs/ethereumjs-monorepo/issues?q=is%3Aopen+is%3Aissue+label%3A"package%3A+trie"
[trie-actions-badge]: https://github.com/ethereumjs/ethereumjs-monorepo/workflows/Trie%20Test/badge.svg
[trie-actions-link]: https://github.com/ethereumjs/ethereumjs-monorepo/actions?query=workflow%3A%22Trie+Test%22
[trie-coverage-badge]: https://codecov.io/gh/ethereumjs/ethereumjs-monorepo/branch/master/graph/badge.svg?flag=trie
[trie-coverage-link]: https://codecov.io/gh/ethereumjs/ethereumjs-monorepo/tree/master/packages/trie


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 4.1.0 - 2021-02-16

This release comes with a reworked checkpointing mechanism (PR [#1030](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1030) and subsequently PR [#1035](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1035)). Instead of copying over the whole DB on checkpoints the operations in between checkpoints are now recorded in memory and either applied in batch on a `Trie.checkpoint()` call or discarded along a `Trie.revert()`. This more fine-grained operational mode leads to a substantial performance gain (up to 50x) when working with larger tries.

Another performance related bug has been fixed along PR [#127](https://github.com/ethereumjs/merkle-patricia-tree/pull/127) removing an unnecessary double-serialization call on nodes. This gives a general performance gain of 10-20% on putting new values in a trie.

Other changes:

## New Features

A new exported `WalkController` class has been added and `trie.walkTrie()` has been made a public method along. This allows for creating own custom ways to traverse a trie. PR [#135](https://github.com/ethereumjs/merkle-patricia-tree/pull/135)

### Refactoring, Development and Documentation

- Better `Trie` code documentation, PR [#125](https://github.com/ethereumjs/merkle-patricia-tree/pull/125)
- Internal `Trie` function reordering & partial retwrite, PR [#125](https://github.com/ethereumjs/merkle-patricia-tree/pull/125)
- Added simple integrated profiling, PR [#128](https://github.com/ethereumjs/merkle-patricia-tree/pull/128)
- Reworked benchmarking to be based on `benchmark.js`, basic CI integration, PR [#130](https://github.com/ethereumjs/merkle-patricia-tree/pull/130)
- Upgrade to `ethereumjs-config` `2.0` libs for linting and formatting, PR [#133](https://github.com/ethereumjs/merkle-patricia-tree/pull/133)
- Switched coverage from `coverall` to `codecov`, PR [#137](https://github.com/ethereumjs/merkle-patricia-tree/pull/137)

## [4.0.0] - 2020-04-17

This release introduces a major API upgrade from callbacks to Promises.

Example using async/await syntax:

```typescript
import { BaseTrie as Trie } from 'merkle-patricia-tree'
const trie = new Trie()
async function test() {
  await trie.put(Buffer.from('test'), Buffer.from('one'))
  const value = await trie.get(Buffer.from('test'))
  console.log(value.toString()) // 'one'
}
test()
```

### Breaking Changes

#### Trie methods

See the [docs](https://github.com/ethereumjs/merkle-patricia-tree/tree/master/docs) for the latest Promise-based method signatures.

#### Trie.prove renamed to Trie.createProof

To clarify the method's purpose `Trie.prove` has been renamed to `Trie.createProof`. `Trie.prove` has been deprecated but will remain as an alias for `Trie.createProof` until removed.

#### Trie raw methods

`getRaw`, `putRaw` and `delRaw` were deprecated in `v3.0.0` and have been removed from this release. Instead, please use `trie.db.get`, `trie.db.put`, and `trie.db.del`. If using a `SecureTrie` or `CheckpointTrie`, use `trie._maindb` to override the checkpointing mechanism and interact directly with the db.

#### SecureTrie.copy

`SecureTrie.copy` now includes checkpoint metadata by default. To maintain original behavior of _not_ copying checkpoint state, pass `false` to param `includeCheckpoints`.

### Changed

- Convert trieNode to ES6 class ([#71](https://github.com/ethereumjs/merkle-patricia-tree/pull/71))
- Merge checkpoint and secure interface with their ES6 classes ([#73](https://github.com/ethereumjs/merkle-patricia-tree/pull/73))
- Extract db-related methods from baseTrie ([#74](https://github.com/ethereumjs/merkle-patricia-tree/pull/74))
- \_lookupNode callback to use standard error, response pattern ([#83](https://github.com/ethereumjs/merkle-patricia-tree/pull/83))
- Accept leveldb in constructor, minor fixes ([#92](https://github.com/ethereumjs/merkle-patricia-tree/pull/92))
- Refactor TrieNode, add levelup types ([#98](https://github.com/ethereumjs/merkle-patricia-tree/pull/98))
- Promisify rest of library ([#107](https://github.com/ethereumjs/merkle-patricia-tree/pull/107))
- Use `Nibbles` type for `number[]` ([#115](https://github.com/ethereumjs/merkle-patricia-tree/pull/115))
- Upgrade sbr-util to 7.0.0 / Upgrade level-mem to 5.0.1 ([#116](https://github.com/ethereumjs/merkle-patricia-tree/pull/116))
- Create dual ES5 and ES2017 builds ([#117](https://github.com/ethereumjs/merkle-patricia-tree/pull/117))
- Include checkpoints by default in SecureTrie.copy ([#119](https://github.com/ethereumjs/merkle-patricia-tree/pull/119))
- Rename Trie.prove to Trie.createProof ([#122](https://github.com/ethereumjs/merkle-patricia-tree/pull/122))

### Added

- Support for proofs of null/absence. Dried up prove/verify. ([#82](https://github.com/ethereumjs/merkle-patricia-tree/pull/82))
- Add more Ethereum state DB focused example accessing account values ([#89](https://github.com/ethereumjs/merkle-patricia-tree/pull/89))

### Fixed

- Drop ethereumjs-testing dep and fix bug in branch value update ([#69](https://github.com/ethereumjs/merkle-patricia-tree/pull/69))
- Fix prove and verifyProof in SecureTrie ([#79](https://github.com/ethereumjs/merkle-patricia-tree/pull/79))
- Fixed src code links in docs ([#93](https://github.com/ethereumjs/merkle-patricia-tree/pull/93))

### Dev / Testing / CI

- Update tape to v4.10.1 ([#81](https://github.com/ethereumjs/merkle-patricia-tree/pull/81))
- Org links and git hooks ([#87](https://github.com/ethereumjs/merkle-patricia-tree/pull/87))
- Use module.exports syntax in util files ([#90](https://github.com/ethereumjs/merkle-patricia-tree/pull/90))

test/index.spec.ts
import tape from 'tape'
import * as rlp from 'rlp'
import { KECCAK256_NULL } from 'sbr-util'
import { BaseTrie, CheckpointTrie } from '../src'

tape('simple save and retrieve', function (tester) {
  const it = tester.test

  it('should not crash if given a non-existent root', async function (t) {
    const root = Buffer.from(
      '3f4399b08efe68945c1cf90ffe85bbe3ce978959da753f9e649f034015b8817d',
      'hex'
    )
    const trie = new CheckpointTrie(null, root)
    const value = await trie.get(Buffer.from('test'))
    t.equal(value, null)
    t.end()
  })

  const trie = new CheckpointTrie()

  it('save a value', async function (t) {
    await trie.put(Buffer.from('test'), Buffer.from('one'))
    t.end()
  })

  it('should get a value', async function (t) {
    const value = await trie.get(Buffer.from('test'))
    t.equal(value!.toString(), 'one')
    t.end()
  })

  it('should update a value', async function (t) {
    await trie.put(Buffer.from('test'), Buffer.from('two'))
    const value = await trie.get(Buffer.from('test'))
    t.equal(value!.toString(), 'two')
    t.end()
  })

  it('should delete a value', async function (t) {
    await trie.del(Buffer.from('test'))
    const value = await trie.get(Buffer.from('test'))
    t.notok(value)
    t.end()
  })

  it('should recreate a value', async function (t) {
    await trie.put(Buffer.from('test'), Buffer.from('one'))
    t.end()
  })

  it('should get updated a value', async function (t) {
    const value = await trie.get(Buffer.from('test'))
    t.equal(value!.toString(), 'one')
    t.end()
  })

  it('should create a branch here', async function (t) {
    await trie.put(Buffer.from('doge'), Buffer.from('coin'))
    t.equal(
      'de8a34a8c1d558682eae1528b47523a483dd8685d6db14b291451a66066bf0fc',
      trie.root.toString('hex')
    )
    t.end()
  })

  it('should get a value that is in a branch', async function (t) {
    const value = await trie.get(Buffer.from('doge'))
    t.equal(value!.toString(), 'coin')
    t.end()
  })

  it('should delete from a branch', async function (t) {
    await trie.del(Buffer.from('doge'))
    const value = await trie.get(Buffer.from('doge'))
    t.equal(value, null)
    t.end()
  })

  tape('storing longer values', async function (tester) {
    const it = tester.test
    const trie = new CheckpointTrie()
    const longString = 'this will be a really really really long value'
    const longStringRoot = 'b173e2db29e79c78963cff5196f8a983fbe0171388972106b114ef7f5c24dfa3'

    it('should store a longer string', async function (t) {
      await trie.put(Buffer.from('done'), Buffer.from(longString))
      await trie.put(Buffer.from('doge'), Buffer.from('coin'))
      t.equal(longStringRoot, trie.root.toString('hex'))
      t.end()
    })

    it('should retrieve a longer value', async function (t) {
      const value = await trie.get(Buffer.from('done'))
      t.equal(value!.toString(), longString)
      t.end()
    })

    it('should when being modified delete the old value', async function (t) {
      await trie.put(Buffer.from('done'), Buffer.from('test'))
      t.end()
    })
  })

  tape('testing extensions and branches', function (tester) {
    const it = tester.test
    const trie = new CheckpointTrie()

    it('should store a value', async function (t) {
      await trie.put(Buffer.from('doge'), Buffer.from('coin'))
      t.end()
    })

    it('should create extension to store this value', async function (t) {
      await trie.put(Buffer.from('do'), Buffer.from('verb'))
      t.equal(
        'f803dfcb7e8f1afd45e88eedb4699a7138d6c07b71243d9ae9bff720c99925f9',
        trie.root.toString('hex')
      )
      t.end()
    })

    it('should store this value under the extension', async function (t) {
      await trie.put(Buffer.from('done'), Buffer.from('finished'))
      t.equal(
        '409cff4d820b394ed3fb1cd4497bdd19ffa68d30ae34157337a7043c94a3e8cb',
        trie.root.toString('hex')
      )
      t.end()
    })
  })

  tape('testing extensions and branches - reverse', function (tester) {
    const it = tester.test
    const trie = new CheckpointTrie()

    it('should create extension to store this value', async function (t) {
      await trie.put(Buffer.from('do'), Buffer.from('verb'))
      t.end()
    })

    it('should store a value', async function (t) {
      await trie.put(Buffer.from('doge'), Buffer.from('coin'))
      t.end()
    })

    it('should store this value under the extension', async function (t) {
      await trie.put(Buffer.from('done'), Buffer.from('finished'))
      t.equal(
        '409cff4d820b394ed3fb1cd4497bdd19ffa68d30ae34157337a7043c94a3e8cb',
        trie.root.toString('hex')
      )
      t.end()
    })
  })
})

tape('testing deletion cases', function (tester) {
  const it = tester.test
  const trie = new CheckpointTrie()

  it('should delete from a branch->branch-branch', async function (t) {
    await trie.put(Buffer.from([11, 11, 11]), Buffer.from('first'))
    await trie.put(Buffer.from([12, 22, 22]), Buffer.from('create the first branch'))
    await trie.put(Buffer.from([12, 34, 44]), Buffer.from('create the last branch'))

    await trie.del(Buffer.from([12, 22, 22]))
    const val = await trie.get(Buffer.from([12, 22, 22]))
    t.equal(null, val)
    t.end()
  })

  it('should delete from a branch->branch-extension', async function (t) {
    await trie.put(Buffer.from([11, 11, 11]), Buffer.from('first'))
    await trie.put(Buffer.from([12, 22, 22]), Buffer.from('create the first branch'))
    await trie.put(Buffer.from([12, 33, 33]), Buffer.from('create the middle branch'))
    await trie.put(Buffer.from([12, 34, 44]), Buffer.from('create the last branch'))

    await trie.del(Buffer.from([12, 22, 22]))
    const val = await trie.get(Buffer.from([12, 22, 22]))
    t.equal(null, val)
    t.end()
  })

  it('should delete from a extension->branch-extension', async function (t) {
    await trie.put(Buffer.from([11, 11, 11]), Buffer.from('first'))
    await trie.put(Buffer.from([12, 22, 22]), Buffer.from('create the first branch'))
    await trie.put(Buffer.from([12, 33, 33]), Buffer.from('create the middle branch'))
    await trie.put(Buffer.from([12, 34, 44]), Buffer.from('create the last branch'))

    // delete the middle branch
    await trie.del(Buffer.from([11, 11, 11]))
    const val = await trie.get(Buffer.from([11, 11, 11]))
    t.equal(null, val)
    t.end()
  })

  it('should delete from a extension->branch-branch', async function (t) {
    await trie.put(Buffer.from([11, 11, 11]), Buffer.from('first'))
    await trie.put(Buffer.from([12, 22, 22]), Buffer.from('create the first branch'))
    await trie.put(Buffer.from([12, 33, 33]), Buffer.from('create the middle branch'))
    await trie.put(Buffer.from([12, 34, 44]), Buffer.from('create the last branch'))
    // delete the middle branch
    await trie.del(Buffer.from([11, 11, 11]))
    const val = await trie.get(Buffer.from([11, 11, 11]))
    t.equal(null, val)
    t.end()
  })
})

tape('shall handle the case of node not found correctly', async (t) => {
  const trie = new BaseTrie()
  await trie.put(Buffer.from('a'), Buffer.from('value1'))
  await trie.put(Buffer.from('aa'), Buffer.from('value2'))
  await trie.put(Buffer.from('aaa'), Buffer.from('value3'))

  /* Setups a trie which consists of 
    ExtensionNode -> 
    BranchNode -> value1
    ExtensionNode -> 
    BranchNode -> value2 
    LeafNode -> value3
  */

  let path = await trie.findPath(Buffer.from('aaa'))

  t.ok(path.node != null, 'findPath should find a node')

  const { stack } = await trie.findPath(Buffer.from('aaa'))
  await trie.db.del(stack[1].hash()) // delete the BranchNode -> value1 from the DB

  path = await trie.findPath(Buffer.from('aaa'))

  t.ok(path.node === null, 'findPath should not return a node now')
  t.ok(
    path.stack.length == 1,
    'findPath should find the first extension node which is still in the DB'
  )

  t.end()
})

tape('it should create the genesis state root from ethereum', function (tester) {
  const it = tester.test
  const trie4 = new CheckpointTrie()

  const g = Buffer.from('8a40bfaa73256b60764c1bf40675a99083efb075', 'hex')
  const j = Buffer.from('e6716f9544a56c530d868e4bfbacb172315bdead', 'hex')
  const v = Buffer.from('1e12515ce3e0f817a4ddef9ca55788a1d66bd2df', 'hex')
  const a = Buffer.from('1a26338f0d905e295fccb71fa9ea849ffa12aaf4', 'hex')

  const stateRoot = Buffer.alloc(32)
  stateRoot.fill(0)

  const startAmount = Buffer.alloc(26)
  startAmount.fill(0)
  startAmount[0] = 1

  const account = [startAmount, 0, stateRoot, KECCAK256_NULL]
  const rlpAccount = rlp.encode(account)
  const cppRlp =
    'f85e9a010000000000000000000000000000000000000000000000000080a00000000000000000000000000000000000000000000000000000000000000000a0c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'

  const genesisStateRoot = '2f4399b08efe68945c1cf90ffe85bbe3ce978959da753f9e649f034015b8817d'
  tester.equal(cppRlp, rlpAccount.toString('hex'))

  it('shall match the root', async function (t) {
    await trie4.put(g, rlpAccount)
    await trie4.put(j, rlpAccount)
    await trie4.put(v, rlpAccount)
    await trie4.put(a, rlpAccount)
    t.equal(trie4.root.toString('hex'), genesisStateRoot)
    t.end()
  })
})


test/encoding.spec.ts
import tape from 'tape'
import { toBuffer } from 'sbr-util'
import { CheckpointTrie } from '../src'

const trie = new CheckpointTrie()
const trie2 = new CheckpointTrie()
const hex = 'FF44A3B3'

tape('encoding hex prefixes', async function (t) {
  await trie.put(Buffer.from(hex, 'hex'), Buffer.from('test'))
  await trie2.put(toBuffer(`0x${hex}`), Buffer.from('test'))
  t.equal(trie.root.toString('hex'), trie2.root.toString('hex'))
  t.end()
})


src/secure.ts
import { keccak256 } from 'sbr-util'
import { CheckpointTrie } from './checkpointTrie'
import { Proof } from './baseTrie'

/**
 * You can create a secure Trie where the keys are automatically hashed
 * using **keccak256** by using `import { SecureTrie as Trie } from 'merkle-patricia-tree'`.
 * It has the same methods and constructor as `Trie`.
 * @class SecureTrie
 * @extends Trie
 * @public
 */
export class SecureTrie extends CheckpointTrie {
  constructor(...args: any) {
    super(...args)
  }

  /**
   * Gets a value given a `key`
   * @param key - the key to search for
   * @returns A Promise that resolves to `Buffer` if a value was found or `null` if no value was found.
   */
  async get(key: Buffer): Promise<Buffer | null> {
    const hash = keccak256(key)
    const value = await super.get(hash)
    return value
  }

  /**
   * Stores a given `value` at the given `key`.
   * For a falsey value, use the original key to avoid double hashing the key.
   * @param key
   * @param value
   */
  async put(key: Buffer, val: Buffer): Promise<void> {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!val || val.toString() === '') {
      await this.del(key)
    } else {
      const hash = keccak256(key)
      await super.put(hash, val)
    }
  }

  /**
   * Deletes a value given a `key`.
   * @param key
   */
  async del(key: Buffer): Promise<void> {
    const hash = keccak256(key)
    await super.del(hash)
  }

  /**
   * prove has been renamed to [[SecureTrie.createProof]].
   * @deprecated
   * @param trie
   * @param key
   */
  static async prove(trie: SecureTrie, key: Buffer): Promise<Proof> {
    return this.createProof(trie, key)
  }

  /**
   * Creates a proof that can be verified using [[SecureTrie.verifyProof]].
   * @param trie
   * @param key
   */
  static createProof(trie: SecureTrie, key: Buffer): Promise<Proof> {
    const hash = keccak256(key)
    return super.createProof(trie, hash)
  }

  /**
   * Verifies a proof.
   * @param rootHash
   * @param key
   * @param proof
   * @throws If proof is found to be invalid.
   * @returns The value from the key.
   */
  static async verifyProof(rootHash: Buffer, key: Buffer, proof: Proof): Promise<Buffer | null> {
    const hash = keccak256(key)
    return super.verifyProof(rootHash, hash, proof)
  }

  /**
   * Returns a copy of the underlying trie with the interface of SecureTrie.
   * @param includeCheckpoints - If true and during a checkpoint, the copy will contain the checkpointing metadata and will use the same scratch as underlying db.
   */
  copy(includeCheckpoints = true): SecureTrie {
    const db = this.db.copy()
    const secureTrie = new SecureTrie(db._leveldb, this.root)
    if (includeCheckpoints && this.isCheckpoint) {
      secureTrie.db.checkpoints = [...this.db.checkpoints]
    }
    return secureTrie
  }
}


dist.browser/trieNode.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRawNode = exports.decodeNode = exports.decodeRawNode = exports.LeafNode = exports.ExtensionNode = exports.BranchNode = void 0;
var rlp = __importStar(require("rlp"));
var sbr_util_1 = require("sbr-util");
var nibbles_1 = require("./util/nibbles");
var hex_1 = require("./util/hex");
var BranchNode = /** @class */ (function () {
    function BranchNode() {
        this._branches = new Array(16).fill(null);
        this._value = null;
    }
    BranchNode.fromArray = function (arr) {
        var node = new BranchNode();
        node._branches = arr.slice(0, 16);
        node._value = arr[16];
        return node;
    };
    Object.defineProperty(BranchNode.prototype, "value", {
        get: function () {
            return this._value && this._value.length > 0 ? this._value : null;
        },
        set: function (v) {
            this._value = v;
        },
        enumerable: false,
        configurable: true
    });
    BranchNode.prototype.setBranch = function (i, v) {
        this._branches[i] = v;
    };
    BranchNode.prototype.raw = function () {
        return __spread(this._branches, [this._value]);
    };
    BranchNode.prototype.serialize = function () {
        return rlp.encode(this.raw());
    };
    BranchNode.prototype.hash = function () {
        return sbr_util_1.keccak256(this.serialize());
    };
    BranchNode.prototype.getBranch = function (i) {
        var b = this._branches[i];
        if (b !== null && b.length > 0) {
            return b;
        }
        else {
            return null;
        }
    };
    BranchNode.prototype.getChildren = function () {
        var children = [];
        for (var i = 0; i < 16; i++) {
            var b = this._branches[i];
            if (b !== null && b.length > 0) {
                children.push([i, b]);
            }
        }
        return children;
    };
    return BranchNode;
}());
exports.BranchNode = BranchNode;
var ExtensionNode = /** @class */ (function () {
    function ExtensionNode(nibbles, value) {
        this._nibbles = nibbles;
        this._value = value;
    }
    ExtensionNode.encodeKey = function (key) {
        return hex_1.addHexPrefix(key, false);
    };
    ExtensionNode.decodeKey = function (key) {
        return hex_1.removeHexPrefix(key);
    };
    Object.defineProperty(ExtensionNode.prototype, "key", {
        get: function () {
            return this._nibbles.slice(0);
        },
        set: function (k) {
            this._nibbles = k;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ExtensionNode.prototype, "value", {
        get: function () {
            return this._value;
        },
        set: function (v) {
            this._value = v;
        },
        enumerable: false,
        configurable: true
    });
    ExtensionNode.prototype.encodedKey = function () {
        return ExtensionNode.encodeKey(this._nibbles.slice(0));
    };
    ExtensionNode.prototype.raw = function () {
        return [nibbles_1.nibblesToBuffer(this.encodedKey()), this._value];
    };
    ExtensionNode.prototype.serialize = function () {
        return rlp.encode(this.raw());
    };
    ExtensionNode.prototype.hash = function () {
        return sbr_util_1.keccak256(this.serialize());
    };
    return ExtensionNode;
}());
exports.ExtensionNode = ExtensionNode;
var LeafNode = /** @class */ (function () {
    function LeafNode(nibbles, value) {
        this._nibbles = nibbles;
        this._value = value;
    }
    LeafNode.encodeKey = function (key) {
        return hex_1.addHexPrefix(key, true);
    };
    LeafNode.decodeKey = function (encodedKey) {
        return hex_1.removeHexPrefix(encodedKey);
    };
    Object.defineProperty(LeafNode.prototype, "key", {
        get: function () {
            return this._nibbles.slice(0);
        },
        set: function (k) {
            this._nibbles = k;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LeafNode.prototype, "value", {
        get: function () {
            return this._value;
        },
        set: function (v) {
            this._value = v;
        },
        enumerable: false,
        configurable: true
    });
    LeafNode.prototype.encodedKey = function () {
        return LeafNode.encodeKey(this._nibbles.slice(0));
    };
    LeafNode.prototype.raw = function () {
        return [nibbles_1.nibblesToBuffer(this.encodedKey()), this._value];
    };
    LeafNode.prototype.serialize = function () {
        return rlp.encode(this.raw());
    };
    LeafNode.prototype.hash = function () {
        return sbr_util_1.keccak256(this.serialize());
    };
    return LeafNode;
}());
exports.LeafNode = LeafNode;
function decodeRawNode(raw) {
    if (raw.length === 17) {
        return BranchNode.fromArray(raw);
    }
    else if (raw.length === 2) {
        var nibbles = nibbles_1.bufferToNibbles(raw[0]);
        if (hex_1.isTerminator(nibbles)) {
            return new LeafNode(LeafNode.decodeKey(nibbles), raw[1]);
        }
        return new ExtensionNode(ExtensionNode.decodeKey(nibbles), raw[1]);
    }
    else {
        throw new Error('Invalid node');
    }
}
exports.decodeRawNode = decodeRawNode;
function decodeNode(raw) {
    var des = rlp.decode(raw);
    if (!Array.isArray(des)) {
        throw new Error('Invalid node');
    }
    return decodeRawNode(des);
}
exports.decodeNode = decodeNode;
function isRawNode(n) {
    return Array.isArray(n) && !Buffer.isBuffer(n);
}
exports.isRawNode = isRawNode;
//# sourceMappingURL=trieNode.js.map